README.md
# PickYourProf

Go to <https://pick-your-prof.herokuapp.com/> to use PickYourProf!

<p><br></p>


<a href="https://ibb.co/ftFGVm"><img src="https://preview.ibb.co/gYZobR/Screen_Shot_2017_10_29_at_11_44_05_AM.png" alt="Screen_Shot_2017_10_29_at_11_44_05_AM" border="0"></a>

<p><br></p>


<a href="https://ibb.co/kHM6Vm"><img src="https://preview.ibb.co/nyfdbR/Screen_Shot_2017_10_29_at_11_44_31_AM.png" alt="Screen_Shot_2017_10_29_at_11_44_31_AM" border="0"></a><br /><a target='_blank' href='https://imgbb.com/'>photo upload sites</a><br />


<h3>Creators</h3>
PickYourProf was made by five CS students, Samarth Desai, Abhinav Kasamsetty, Sandeep Namburi, Joel Uong, and Rithvik Vellaturi, at HackTX 2017.

<h3>Inspiration</h3>
There was no easy way to get personalized professor choices. Looking for a professor requires hours of looking through numerous review sites, university catalogs, etc. We've simplified that for you.

<h3>What it does</h3>
Pick My Prof analayzes the UT directory of professors and gives your recommendations for professors based on Rate My Professor ratings and university reviews. It currently works for every single course and professor in the UT Austin database.

<h3>How we built it</h3>
Our app has a light front-end with HTML and CSS. We have a powerful node web app powered by an Express.js server. The Node.js file is responsible for scraping data from websites and accessing a database file that lists every UT professor and course. We utilized the Fetch API to pull and push data between scripts.

<h3>How it works</h3>
Once the user enters a course, the program gets the list of professors who teach the course and the distribution of grades
for each professor for the given course. Then, it scrapes RateMyProfessors in order to get the overall quality, difficulty, average sentiment of comments, and percentage of students who would take a course by the professor again for each of the professors in the list.

With this data, the algorithm normalizes the overall quality, difficulty, average sentiment, and take again percentage to be out of 5, with 5 being the best. The algorithm then does a weight average of the metrics and adds it to the average GPA of a student taking a course with the professor, to get the PickYourProf score, which is on a scale from 0 to 10.

<h3>How to open source</h3>
We've included the bootsrap folder we've implemented material design. algorithm.js, server.js, and index.html are the main source dependencies. To add to our algorithm or frontend and expand to other schools, user the files listed above.


CHANGELOG.md
# Changelog for PickYourProf

## [2.3.1] - 2019-10-23
### Fixed
- Bug where checking if table row ID was a number didn't handle empty strings

## [2.3.0] - 2018-11-02
### Added
- Caching of Rate My Professors data

### Changed
- Updated database with 2018 data

## [2.2.0] - 2018-06-29
### Added
- Took sentiment analysis of comments into account in the algorithm
- Added how the algorithm works

### Changed
- Made RMP metrics other than overall quality worth less

## [2.1.0] - 2018-05-31
### Added
- Scraped difficulty rating and take again percentage metrics from Rate My Professors to include in algorithm calculations
- This changelog

### Fixed
- Style fixes

## [2.0.3] - 2018-04-24
### Fixed
- Undefined score when none of the professors were found on Rate My Professors

## [2.0.2] - 2018-04-23
### Added
- Site displays scores along with names in table format for the results
- Creator names to README

### Changed
- Rewrote backend with promises instead of callbacks
- User input for course department and number made case insensitive
- Made algorithm treat A and A+ students separately
- Cleaned up code for calculating percentages style wise

### Fixed
- Handled case where prof has no rating on Rate My Professors page

## [2.0.1] - 2018-04-05
### Changed
- Reorganized files in the repository

## [2.0.0] - 2018-04-05
### Changed
- UT grade data reliance switched from JSON to SQL database
- README updates

### Fixed
- Style fixes

### Removed
- Unnecessary files

## [1.0.0] - 2017-10-29
### Added
- Initial release at HackTX 2017

scraper.js
var google = require('google')

google.resultsPerPage = 25
var nextCounter = 0

var prompt = require('prompt');
var casper = require('casper').create();
var opn = require('opn');


prompt.get(['firstName2', 'lastName2'], function (err, result) {
    if (err) { return onErr(err); }
    console.log('Command-line input received:');
    console.log('  First Name: ' + result.firstName2);
    console.log('  Last Name: ' + result.lastName2);

    var URL = "http://www.ratemyprofessors.com/search.jsp?queryoption=HEADER&queryBy=teacherName&schoolName=University+of+Texas+at+Austin&schoolID=1255&query="+result.firstName+"+"+result.lastName;

    google(result.firstName2 + result.lastName2 + 'university of texas rate my professor', function (err, res)  {
      if (err) console.error(err)

        var link = res.links[0];
        console.log(link.href)

        var request = require("request"),
          cheerio = require("cheerio"),
          url = link.href;

        request(url, function (error, response, body) {
          if (!error) {
            var $ = cheerio.load(body),
              firstName = $("[class='pfname']").html();
              lastName = $("[class='plname']").html();
              difficulty = $("[class='breakdown-section difficulty'] .grade").html();
              overallQuality = $("[class='breakdown-container quality'] .grade").html();
              var firstName = firstName.replace(/\s+/g, "");
              var lastName = lastName.replace(/\s+/g, "");
              var difficulty = difficulty.replace(/\s+/g, "");
              var overallQuality = overallQuality.replace(/\s+/g, "");

            console.log(firstName + " " + lastName + "'s overall difficulty is " + difficulty + ".");
            console.log(firstName + " " + lastName + "'s overall quality is " + overallQuality + ".");

          } else {
            console.log("Sorry we messed up somewhere: " + error);
          }
        });

    })

  });

  function onErr(err) {
    console.log(err);
    return 1;
  }

  $('.btn').click(function() {

    $('.text').text('loading . . .');

    $.ajax({

      type:"GET",
      url:"https://api.meetup.com/2/cities",
      success: function(data) {
        $('.text').text(JSON.stringify(data));
      },
      dataType: 'jsonp',
      error: function() {
        console.log("Something went wrong, data could not be fetched");
      }
    });
    console.log('hello', hello);
    //console.log($("#city").text());
  });


src/server.js
const getProfData = require('./algorithm');
const express = require('express');
const bodyParser = require('body-parser');
const path = require('path');
const multer = require('multer');

const app = express();
const upload = multer();

let PORT = process.env.PORT || 3000;

app.use(bodyParser.json()); // for parsing application/json
app.use(bodyParser.urlencoded({ extended: true }));
app.use(express.static(path.join(__dirname, '..', 'public')));

app.post('/home', upload.any(), (req, res) => {
  let course = req.body.course;
  console.log('Course:', course);

  // Get the sorted list of professors and their scores
  getProfData(course)
    .then(data => {
      if (data.list.length == 0)
        res.send('Not found in database. Ensure spelling.');
      else {
        // HTML tags for output formatting
        const tag = '<td style="font-size:18px;font-family:Roboto;color:black;">';
        let output = '<table style="width:40%;margin-left:26%;margin-right:20%;">';
        output += `<tr>${tag}<b>Professor</b></td>${tag}<b>Score</b></td></tr>`;
        for (let prof of data.list) {
          // Formats the name correctly
          let name = formatFullName(prof);
          output += '<tr>';
          output += (tag + name + '</td>');
          output += (tag + (Math.round(data.scores[prof] * 100) / 100) + '</td>');
          output += '</tr>';
        }
        output += '</table>';
        // Sends output to HTML file to be printed on the webpage
        res.send(output);
      }
    });

});

app.get('/', (req, res) => {
  res.sendFile(path.join(__dirname, '..', 'public', 'index.html'));
});

app.listen(PORT, () => {
  console.log('Example app listening on http://localhost:' + PORT);
});

function formatFullName(name) {
  // Capitalize only the first letter of first and last name and account for hyphens
  let tokens = name.split(' ');
  let first = formatToken(tokens[0]);
  let last = formatToken(tokens[1]);
  return first + ' ' + last;

  function formatToken(token) {
    token = token.substring(0, 1) + token.substring(1).toLowerCase();
    let arr = token.split('-');
    return arr.length < 2 ? token : `${arr[0]}-${arr[1].substr(0, 1).toUpperCase()}${arr[1].substr(1)}`;
  }
}


catalystFull.js
/**
 *
 */
const Request = require("request");
const Cheerio = require("cheerio");
const sqlite3 = require('sqlite3').verbose();

module.exports = function myFunc(courseNumber) {
  return new Promise(function(resolve, reject) {
    let db = new sqlite3.Database("./grades.db", sqlite3.OPEN_READONLY, (err) => {
      if (err) {
        console.error(err.message);
      }
      console.log("Connected to the grades database.");
    });
    
    var rts = courseNumber.split(" ");
    db.serialize(() => {
      let sql = `SELECT prof AS name, a1, a2, a3, b1, b2, b3, c1, c2, c3, d1, d2, d3, f
                FROM agg
                WHERE dept = ? AND course_nbr = ?
                ORDER BY name`;

      var profs = [];
      var percentages = [];
      db.each(sql, [rts[0], rts[1]], (err, row) => {
        if (err) {
          throw err;
        }
        var splitName = row.name.split(/ +/);
        var finalName = splitName[1] + " " + splitName[0];
        profs.push(finalName);

        var totalStudents = row.a1 + row.a2 + row.a3 + row.b1 + row.b2 + row.b3
                          + row.c1 + row.c2 + row.c3 + row.d1 + row.d2 + row.d3
                          + row.f;
        var currPercentages = [];
        currPercentages.push((row.a1 + row.a2) / totalStudents * 100);
        currPercentages.push(row.a3 / totalStudents * 100);
        var letters = ["b", "c", "d"];
        var numbers = ["1", "2", "3"];
        letters.forEach((letter) => {
          numbers.forEach((number) => {
            currPercentages.push(row[letter + number] / totalStudents * 100);
          });
        });
        currPercentages.push(row.f / totalStudents * 100);
        percentages.push(currPercentages);
      }, () => {
        // console.log(profs);
        // console.log(percentages);

        ratings(profs, percentages, function(res) {
          var resCopy = res.concat([]);
          console.log(resCopy);
          resCopy.sort();
          var profsCopy = [];
          for (var i = resCopy.length - 1; i > -1; i--) {
            var index = res.indexOf(resCopy[i]);
            profsCopy.push(profs[index]);
          }
          console.log(profsCopy);
          // FINAL OUTPUT
          resolve(profsCopy);
        });
      });
      
      db.close();
    });

    function ratings(array, percentages, callback) {
      overallRating(array, function(res) {
        var sum = 0.0;
        for (var i = 0; i < res.length; i++) {
          if (res[i] != -1)
            sum += res[i];
        }
        var avg = sum / res.length;

        var scores = [];

        var index = -1;
        var best = -0.1;
        for (var i = 0; i < array.length; i++) {
          if (res[i] === -1) {
            res[i] = avg;
          }
          var gpa = [4, 3.67, 3.33, 3, 2.67, 2.33, 2, 1.67, 1.33, 1, 0.67, 0];
          var avgGpa = 0;
          for (var j = 0; j < percentages[i].length; j++) {
            avgGpa += percentages[i][j] * gpa[j];
          }
          avgGpa /= 100.0;
          avgGpa = avgGpa * 5.0 / 4.0;

          var score = avgGpa + res[i];
          scores.push(score);
        }

        callback(scores);

      });
    }

    //Gets an array of ratings
    function overallRating(array, callback) {
      var json = {};
      // Goes through list of professors
      for (var i = 0; i < array.length; i++) {
        var name = array[i];
        name = name.replace(" ", "+");
        // Link to query Rate My Professors
        var link = "http://www.ratemyprofessors.com/search.jsp?queryoption=HEADER&queryBy=teacherName" +
          "&schoolName=University+of+Texas+at+Austin&schoolID=1255&query=" + name;

        // Queries the link to the professor search results
        Request(link, function(err, resp, body) {
          if (!err && resp.statusCode == 200) {
            var $ = Cheerio.load(body);

            // Checks if any results are found
            var htmas = $("div.not-found-box").next().html();
            if (!htmas.includes("Your search")) {
              // Gets the link for the first result
              $(".listings").filter(function() {
                var endLink = $($(this)[0].children[3].children[1]).attr("href");
                var link2 = "http://www.ratemyprofessors.com" + endLink;

                // Queries the link to the professor's page
                Request(link2, function(error, response, html) {
                  if (!error && response.statusCode == 200) {
                    var $ = Cheerio.load(html),
                    overallQuality = $("[class='breakdown-container quality'] .grade").html();
                    if (overallQuality != null) {
                      var overallQuality = overallQuality.replace(/\s+/g, "");
                      firstName = $(".pfname").html();
                      lastName = $(".plname").html();
                      var firstName = firstName.replace(/\s+/g, "");
                      var lastName = lastName.replace(/\s+/g, "");
                      //console.log(firstName + " " + lastName + " " + overallQuality);
                      json[lastName.toUpperCase()] = overallQuality;
                    } else {
                      var overallQuality = -1.0;
                      name = $("div.name").html();
                      //console.log(name + " " + overallQuality);
                      var i = name.indexOf(" ");
                      json[name.substring(i + 1).toUpperCase()] = overallQuality;
                    }
                  }
                });
              });
            } else {
              // No results found in Rate My Professors
              //console.log("Doesn't matter" + " " + -1.0);
            }
          }
        });
      }

      // Converts json to array after 2.5 seconds
      setTimeout(function() {
        var arr = [];
        for (var i = 0; i < array.length; i++) {
          var index = array[i].indexOf(" ");
          if (json[array[i].substring(index + 1)] == null)
            arr[i] = -1;
          else
            arr[i] = parseFloat(json[array[i].substring(index + 1)]);
        }
        callback(arr);
      }, 2500);
    }
  });
}


src/algorithm.js
/**
 *  algorithm.js
 *  Takes in the name of a course and ranks professors that teach the course
 */
const fetch = require('node-fetch');
const Cheerio = require('cheerio');
const sqlite3 = require('sqlite3').verbose();
const Sentiment = require('sentiment');
const sentiment = new Sentiment();
const moment = require('moment');

// SQL query for getting professors and their grade data for a course
let PROF_QUERY = `SELECT prof AS name, a1, a2, a3, b1, b2, b3, c1, c2, c3, d1, d2,
                  d3, f, rating, difficulty, take_again, sentiment, last_updated
                  FROM agg
                  INNER JOIN rmp_cache
                  ON agg.prof=rmp_cache.professor
                  WHERE dept = ? AND course_nbr = ?
                  ORDER BY name`;

// SQL query for caching new RMP data
let CACHE_QUERY = `UPDATE rmp_cache
                  SET rating = ?, difficulty = ?, take_again = ?, sentiment = ?, last_updated = ?
                  WHERE professor = ?;`

module.exports = courseNumber => {
  return new Promise((resolve, reject) => {
    let db = new sqlite3.Database('./grades.db', sqlite3.OPEN_READWRITE, err => {
      if (err) {
        console.error(err.message);
      } else {
        console.log('Connected to the grades database.');
      }
    });

    let rts = courseNumber.split(' ');
    db.serialize(() => {
      // List of professor objects
      let profs = [];
      // Gets information on each professor teaching the given course
      db.each(PROF_QUERY, rts, (err, row) => {
        if (err) {
          throw err;
        }
        // Reorders full name into first name, last name
        let splitName = row.name.split(/,?\s+/);
        let finalName = splitName[1] + ' ' + splitName[0];
        row.oldName = row.name;
        row.name = finalName;
        let currPerc = getGradePercentages(row.a1, row.a2, row.a3, row.b1, row.b2,
          row.b3, row.c1, row.c2, row.c3, row.d1, row.d2, row.d3, row.f);
        row.currPerc = currPerc;
        row.dirty = false;
        profs.push(row);
      }, () => {
        let promises = profs.map(prof => getProfInfo(prof));
        // Gets the Rate My Professors info for each professor
        // Calculates the Pick Your Prof score for each professor
        Promise.all(promises)
          .then(responses => {
            profs.forEach((prof, i) => {
              let resp = responses[i];
              // Cache new RMP data to database if prof object is dirty
              if (prof.dirty) {
                db.run(CACHE_QUERY, resp.rating, resp.difficulty, resp.again, resp.sentimentAvg, prof.last_updated, prof.oldName);
              }
            });

            let scores = getScores(profs, responses);
            // Zip up the profs and scores into an object
            let profData = {};
            profs.forEach((prof, i) => profData[prof.name] = scores[i]);
            console.log(profData);
            // Sort professors by their scores in descending order
            let sortedProfs = Object.keys(profData).sort((a, b) => {
              return profData[b] - profData[a];
            });

            // Wrap sorted professor list and score data into an object
            let finalData = {
              list: sortedProfs,
              scores: profData
            };
            db.close();
            // FINAL OUTPUT
            resolve(finalData);
          });
      });
    });
  });
}

// Stores info of the percentage of each class that got each grade
function getGradePercentages(...distribution) {
  let totalStu = distribution.reduce((a, b) => a + b, 0);
  distribution.forEach((numStu, i) => distribution[i] = numStu / totalStu);
  return distribution;
}

// Gets the PickYourProf score for each professor
function getScores(profs, responses) {
  // Gets the average of all the professor's metric averages
  let numProfs = 0;
  let sum = 0;
  let metricAverages = [];
  let metricAverage;
  for (let resp of responses) {
    if (!resp.rating && !resp.difficulty && !resp.again && !resp.sentimentAvg) {
      metricAverages.push(-1);
    } else {
      numProfs++;
      metricAverage = getMetricAverage(resp);
      metricAverages.push(metricAverage);
      sum += metricAverage;
    }
  }
  let avg = sum / numProfs;

  let scores = [];
  let score, avgGpa;
  for (let i = 0; i < profs.length; i++) {
    // Sets Rate My Professor metric average if professor has no metrics
    if (metricAverages[i] === -1)
      metricAverages[i] = avg;

    // GPA that a student gets for getting between an A+ and an F
    const gpa = [4, 4, 3.67, 3.33, 3, 2.67, 2.33, 2, 1.67, 1.33, 1, 0.67, 0];
    // Calculates average GPA of students in the course
    avgGpa = 0;
    profs[i].currPerc.forEach((percent, j) => avgGpa += percent * gpa[j]);
    // Normalize average GPA to be out of 5 instead of 4
    avgGpa *= 5 / 4;

    // Score is normalized average GPA + Rate My Professors metric average
    // If none of the professors are on RMP, score is avg GPA scaled out of 10
    // Max score is 10
    score = numProfs !== 0 ? avgGpa + metricAverages[i] : avgGpa * 2;
    scores.push(score);
  }
  return scores;

  // Calculates average of the three metrics for each professor
  function getMetricAverage(res) {
    let metricSum = 0;
    let numMetrics = 0;
    for (let prop in res) {
      // Make rating worth more than sentiment, difficulty and take again
      if (res[prop]) {
        if (prop === 'rating') {
          metricSum += res[prop];
          numMetrics++;
        } else {
          metricSum += res[prop] / 4;
          numMetrics += 1 / 4;
        }
      }
    }
    return metricSum / numMetrics;
  }
}

// Base URL for HTTP request
const baseURL = 'http://www.ratemyprofessors.com/search.jsp?queryoption=HEADER' +
  '&queryBy=teacherName&schoolName=University+of+Texas+at+Austin&schoolID=1255&query=';

// Gets the rating for a professor
function getProfInfo(prof) {
  return new Promise((resolve, reject) => {
    // If the cache entry is empty or old
    if (!prof.last_updated || moment().diff(moment(prof.last_updated), 'days') > 7) {
      // Set dirty to true if need to scrape RateMyProfessors for new data
      prof.dirty = true;
      prof.last_updated = moment().valueOf();
      let name = prof.name.replace(' ', '+');
      let link = baseURL + name;
      // Searches for the professor on Rate My Professors
      // If professor found, then gets the professor's rating
      scrapeProfLink(link)
        .then(endLink => {
          scrapeProfInfo('http://www.ratemyprofessors.com' + endLink)
            .then(resp => {
              normalizeMetrics(resp);
              return resolve(resp);
            })
            .catch(e => resolve(e));
        })
        .catch(err => resolve(err));
    } else {
      // If the cache entry is less than a week old
      resolve({
        rating: prof.rating,
        difficulty: prof.difficulty,
        again: prof.take_again,
        sentimentAvg: prof.sentiment
      });
    }
  });

  // Normalize metrics in response to be out of 5.0 and fix types
  function normalizeMetrics(res) {
    res.rating = res.rating ? parseFloat(res.rating) : undefined;
    // 5 - difficulty because we want to make higher difficulty be lower rating
    res.difficulty = res.difficulty ? 5 - parseFloat(res.difficulty) : undefined;
    // Again percentage / 20 to normalize it out of 5
    res.again = res.again ? parseFloat(res.again) / 20 : undefined;
    // Change sentiment scale from -5 to 5 to 0 to 5
    res.sentimentAvg = res.sentimentAvg ? (res.sentimentAvg + 5) / 2 : undefined;
  }
}

// Scrapes link of professor on Rate My Professors
function scrapeProfLink(link) {
  return new Promise((resolve, reject) => {
    fetch(link)
      .then(res => res.text())
      .then(body => {
        let $ = Cheerio.load(body);
        // Checks if any results are found
        let htmas = $('div.not-found-box').next().html();
        if (!htmas.includes('Your search')) {
          // Gets the link for the first result
          $('.listings').filter(function () {
            let endLink = $($(this)[0].children[3].children[1]).attr('href');
            resolve(endLink);
          });
        } else {
          // No results found in Rate My Professors for the professor
          reject({});
        }
      })
      .catch(err => reject({}));
  });
}

// Possible keys in the response object when getting professor info
const responseKeys = ['rating', 'difficulty', 'again', 'sentimentAvg'];

// Scrapes info of professor on Rate My Professors
function scrapeProfInfo(link) {
  return new Promise((resolve, reject) => {
    fetch(link)
      .then(resp => resp.text())
      .then(html => {
        // Queries the link to the professor's page
        let $ = Cheerio.load(html);
        let response = {};
        let rating = $('div.breakdown-container.quality .grade').html();
        if (rating) {
          rating = rating.replace(/\s+/g, '');
          // Overall rating found for professor
          response.rating = rating === 'N/A' ? undefined : rating;
        }
        let difficulty = $('div.breakdown-section.difficulty .grade').html();
        if (difficulty) {
          difficulty = difficulty.replace(/\s+/g, '');
          // Difficulty found for professor
          response.difficulty = difficulty === 'N/A' ? undefined : difficulty;
        }
        let again = $('div.breakdown-section.takeAgain .grade').html();
        if (again) {
          again = again.replace(/[\s%]+/g, '');
          // Take again percentage found for professor
          response.again = again === 'N/A' ? undefined : again;
        }
        let numComments = 0;
        let sentimentSum = 0;
        let comment, sentimentRes, commentSentiment, id;
        $('tr').each(function (i, elem) {
          id = $(this).attr('id');
          // Checks if id exists, has a nonzero length, and is a number
          if (id && !isNaN(id)) {
            comment = $(this).children()[2].children[3].children[0].data.trim();
            sentimentRes = sentiment.analyze(comment);
            if (sentimentRes.words.length === 0)
              commentSentiment = 0;
            else
              commentSentiment = sentimentRes.score / sentimentRes.words.length;
            sentimentSum += commentSentiment;
            numComments++;
          }
        });
        response.sentimentAvg = numComments === 0 ? undefined : sentimentSum / numComments;
        // Reject if response has undefined values for each 
        if (!(response.rating ||
          response.difficulty ||
          response.again ||
          response.sentimentAvg)) {
          reject({});
        }
        resolve(response);
      });
  });
}
